### PR TITLE
typing scheduler.py [2/2]: Apply types

### PR DESCRIPTION
**Pull request by aorenste**
Add `# mypy: disallow-untyped-defs` to scheduler.py and then fix the resulting fallout.

We probably should eventually add a new node between BaseSchedulerNode and all the non-FusedSchedulerNode types to indicate the split between nodes that have a valid `self.node` and ones that don't. That would cause a lot of the `assert self.node is not None` churn to go away - but was a bigger change because a lot of code makes assumptions about types that aren't reflected in the types themselves.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126656

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by eellison — APPROVED** [review]
nice, looks good! could you rebase onto more stable master to land ?

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -691,15 +719,17 @@ class NopKernelSchedulerNode(BaseSchedulerNode):
...
+    assert multi_template is None or isinstance(multi_template, ir.MultiTemplateBuffer)
+    if multi_template and multi_template.make_kernel_render is None:
         lines.append(f"{node.get_name()} Unfinalized multi template buffer")
     else:
+        from torch._inductor.codegen.triton import TritonScheduling
+
         snodes = (node,) if isinstance(node, SchedulerNode) else node.snodes
         device = snodes[0].get_device()
         backend = node.scheduler.get_backend(device)
+        assert isinstance(backend, TritonScheduling)
```
Comment (shunting314) [inline torch/_inductor/scheduler.py:732]: This assert fails. Not sure why CI does not catch it.

Comment (shunting314) [reply torch/_inductor/scheduler.py:732]: oh, maybe CI caches it, but this is caught and skipped

Comment (Chillee) [reply torch/_inductor/scheduler.py:732]: I ran into this as well.

Comment (aorenste) [reply torch/_inductor/scheduler.py:732]: Fixed in #127021

**Pull request closed without merging**